### PR TITLE
zh-cn: update iterators_and_generators

### PR DESCRIPTION
**Pull request by aquanjsw**
### Description

- Sync with upstream
- Fix flaws

### CODE REVIEW

File: files/zh-cn/web/javascript/guide/iterators_and_generators/index.md [hunk 1]
```diff
@@ -16,21 +16,30 @@ slug: Web/JavaScript/Guide/Iterators_and_generators
...
-Javascript 中最常见的迭代器是 Array 迭代器，它只是按顺序返回关联数组中的每个值。虽然很容易想象所有迭代器都可以表示为数组，但事实并非如此。数组必须完整分配，但迭代器仅在必要时使用，因此可以表示无限大小的序列，例如 0 和无穷大之间的整数范围。
+- `value`
+  - : 迭代序列的下一个值。
+- `done`
+  - : 如果已经迭代到序列中的最后一个值，则它为 `true`。如果 `value` 和 `done` 一起出现，则它就是迭代器的返回值。
 
-这是一个可以做到这一点的例子。它允许创建一个简单的范围迭代器，它定义了从开始（包括）到结束（独占）间隔步长的整数序列。它的最终返回值是它创建的序列的大小，由变量 iterationCount 跟踪。
+一旦创建，迭代器对象可以通过重复调用 `next()` 显式地迭代。迭代一个迭代器被称为消耗了这个迭代器，因为它通常只能执行一次：在产生终值后，对 `next()` 的额外调用应该继续返回 `{done：true}`。
+
+Javascript 中最常见的迭代器是 Array 迭代器，它按顺序返回关联数组中的每个值。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/guide/iterators_and_generators/index.md:30]:
```suggestion
Javascript 中最常见的迭代器是数组迭代器，它按顺序返回关联数组中的每个值。
```

File: files/zh-cn/web/javascript/guide/iterators_and_generators/index.md [hunk 2]
```diff
@@ -56,41 +65,68 @@ while (!result.done) {
   result = it.next();
 }
 
-console.log("Iterated over sequence of size: ", result.value); // 5
+console.log(`已迭代序列的大小：${result.value}`); // 5
 ```
 
-> **备注：** 反射性地知道特定对象是否是迭代器是不可能的。如果你需要这样做，请使用 [Iterables](#Iterables).
+> **备注**：[反射性](https://zh.wikipedia.org/wiki/反射式编程)地知道特定对象是否是迭代器是不可能的。如果你需要这样做，请使用[可迭代对象](#可迭代对象).
```
Comment (yin1999) [inline files/zh-cn/web/javascript/guide/iterators_and_generators/index.md:71]: Please reserve this format, or the block will not be highlighted

image

```suggestion
> **备注：** [反射性](https://zh.wikipedia.org/wiki/反射式编程)地知道特定对象是否是迭代器是不可能的。如果你需要这样做，请使用[可迭代对象](#可迭代对象)。
```

File: files/zh-cn/web/javascript/guide/iterators_and_generators/index.md [hunk 3]
```diff
@@ -56,41 +65,68 @@ while (!result.done) {
...
 
 ## 生成器函数
 
-虽然自定义的迭代器是一个有用的工具，但由于需要显式地维护其内部状态，因此需要谨慎地创建。生成器函数提供了一个强大的选择：它允许你定义一个包含自有迭代算法的函数，同时它可以自动维护自己的状态。生成器函数使用 {{jsxref("Statements/function*","function*")}}语法编写。最初调用时，生成器函数不执行任何代码，而是返回一种称为 Generator 的迭代器。通过调用生成器的下一个方法消耗值时，Generator 函数将执行，直到遇到 yield 关键字。
+虽然自定义迭代器是一个有用的工具，但由于需要显式地维护其内部状态，因此创建时要格外谨慎。**生成器函数**（Generator 函数）提供了一个强大的替代选择：它允许你定义一个非连续执行的函数作为迭代算法。生成器函数使用 {{jsxref("Statements/function*","function*")}} 语法编写。
+
+最初调用时，生成器函数不执行任何代码，而是返回一种称为 **Generator** 的特殊迭代器。通过调用 `next()` 方法消耗该生成器时，Generator 函数将执行，直至遇到 `yield` 关键字。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/guide/iterators_and_generators/index.md:77]:
```suggestion
最初调用时，生成器函数不执行任何代码，而是返回一种称为**生成器**的特殊迭代器。通过调用 `next()` 方法消耗该生成器时，生成器函数将执行，直至遇到 `yield` 关键字。
```

File: files/zh-cn/web/javascript/guide/iterators_and_generators/index.md [hunk 4]
```diff
@@ -56,41 +65,68 @@ while (!result.done) {
...
 
 可以根据需要多次调用该函数，并且每次都返回一个新的 Generator，但每个 Generator 只能迭代一次。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/guide/iterators_and_generators/index.md:79]:
```suggestion
可以根据需要多次调用该函数，并且每次都返回一个新的生成器，但每个生成器只能迭代一次。
```

File: files/zh-cn/web/javascript/guide/iterators_and_generators/index.md [hunk 5]
```diff
@@ -123,7 +161,7 @@ for (let value of myIterable) {
 
 ### 用于可迭代对象的语法
 
-一些语句和表达式专用于可迭代对象，例如 {{jsxref("Statements/for...of","for-of")}} 循环，{{jsxref("Operators/Spread_operator","展开语法")}}，{{jsxref("Operators/yield*", "yield*")}} 和 {{jsxref("Operators/Destructuring_assignment", "解构赋值")}}。
+一些语句和表达式专用于可迭代对象，例如 {{jsxref("Statements/for...of", "for...of")}} 循环，{{jsxref("Operators/Spread_syntax", "展开语法")}}，{{jsxref("Operators/yield*", "yield*")}} 和 {{jsxref("Operators/Destructuring_assignment", "解构赋值")}}。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/guide/iterators_and_generators/index.md:164]: Could you update the en-US document to reflect the following translation?

```suggestion
一些语句和表达式专用于可迭代对象，例如 {{jsxref("Statements/for...of", "for...of")}} 循环、{{jsxref("Operators/Spread_syntax", "展开语法", "", 1)}}、{{jsxref("Operators/yield*", "yield*")}} 和{{jsxref("Operators/Destructuring_assignment", "解构赋值", "", 1)}}。
```

File: files/zh-cn/web/javascript/guide/iterators_and_generators/index.md [hunk 6]
```diff
@@ -183,10 +221,10 @@ console.log(sequence.next().value); // 1
...
 ```
 
-你可以通过调用其 {{jsxref("Global_Objects/Generator/throw","throw()")}} 方法强制生成器抛出异常，并传递应该抛出的异常值。这个异常将从当前挂起的生成器的上下文中抛出，就好像当前挂起的 `yield` 是一个 `throw value` 语句。
+你可以通过调用其 {{jsxref("Global_Objects/Generator/throw", "throw()")}} 方法强制生成器抛出异常，并传递应该抛出的异常值。这个异常将从当前挂起的生成器的上下文中抛出，就好像当前挂起的 `yield` 是一个 `throw value` 语句。
 
-如果在抛出的异常处理期间没有遇到 `yield`，则异常将通过调用 `throw()` 向上传播，对 `next()` 的后续调用将导致 `done` 属性为 `true`。
+如果该异常没有在生成器内部被捕获，则它将通过 `throw()` 的调用向上传播，对 `next()` 的后续调用将导致 `done` 属性为 `true`。
 
-生成器具有 {{jsxref("Global_Objects/Generator/return","return(value)")}} 方法，返回给定的值并完成生成器本身。
+生成器的 {{jsxref("Global_Objects/Generator/return", "return()")}} 方法可返回给定的值并终结这个生成器。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/guide/iterators_and_generators/index.md:228]:
```suggestion
生成器的 {{jsxref("Generator/return", "return()")}} 方法可返回给定的值并终结这个生成器。
```

File: files/zh-cn/web/javascript/guide/iterators_and_generators/index.md [hunk 7]
```diff
@@ -147,29 +185,29 @@ a; // "a"
 
 ## 高级生成器
 
-生成器会按需计算它们的产生值，这使得它们能够有效的表示一个计算成本很高的序列，甚至是如上所示的一个无限序列。
+生成器会*按需*计算它们 `yield` 的值，这使得它们能够高效地表示一个计算成本很高的序列，甚至是前文所示的一个无限序列。
+
+{{jsxref("Global_Objects/Generator/next", "next()")}} 方法也接受一个参数用于修改生成器内部状态。传递给 `next()` 的参数值会被 `yield` 接收。
 
-The {{jsxref("Global_Objects/Generator/next","next()")}} 方法也接受一个参数用于修改生成器内部状态。传递给 `next()` 的参数值会被 yield 接收。要注意的是，传给第一个 `next()` 的值会被忽略。
+> **备注**：传给*第一个* `next()` 的值会被忽略。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/guide/iterators_and_generators/index.md:192]:
```suggestion
> **备注：** 传给*第一个* `next()` 的值会被忽略。
```

File: files/zh-cn/web/javascript/guide/iterators_and_generators/index.md (hunk 5 above)
Comment (AquanJSW) [reply files/zh-cn/web/javascript/guide/iterators_and_generators/index.md]: https://github.com/mdn/content/pull/30054

File: files/zh-cn/web/javascript/guide/iterators_and_generators/index.md [hunk 8]
```diff
@@ -123,7 +161,7 @@ for (let value of myIterable) {
 
 ### 用于可迭代对象的语法
 
-一些语句和表达式专用于可迭代对象，例如 {{jsxref("Statements/for...of","for-of")}} 循环，{{jsxref("Operators/Spread_operator","展开语法")}}，{{jsxref("Operators/yield*", "yield*")}} 和 {{jsxref("Operators/Destructuring_assignment", "解构赋值")}}。
+一些语句和表达式专用于可迭代对象，例如 {{jsxref("Statements/for...of", "for...of")}} 循环、{{jsxref("Operators/Spread_syntax", "展开语法", "", 1)}}、{{jsxref("Operators/yield*", "yield*")}} 和{{jsxref("Operators/Destructuring_assignment", "解构赋值", "", 1)}}。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/guide/iterators_and_generators/index.md:164]:
```suggestion
一些语句和表达式专用于可迭代对象，例如 {{jsxref("Statements/for...of", "for...of")}} 循环、{{jsxref("Operators/Spread_syntax", "展开语法", "", 1)}}、{{jsxref("Operators/yield*", "yield*")}} 和{{jsxref("Operators/Destructuring_assignment", "解构", "", 1)}}语法。
```

**Review by yin1999 — APPROVED** [review]
LGTM

**Pull request merged**